File: tests/xcalablemp/NOT-IMPLEMENTED/288/testp154.f90
! testp154.f
! loop指示文とreduction節 (-) のテスト

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N)
!$xmp distribute t(cyclic) onto p
      integer a(N), sa
      real*4 b(N), sb
      real*8 c(N), sc
      integer,allocatable:: w(:)
      integer ans, procs
!$xmp align a(i) with t(i)
!$xmp align b(i) with t(i)
!$xmp align c(i) with t(i)
      character(len=2) result

!$xmp loop (i) onto t(i)
      do i=1, N
         a(i) = xmp_node_num()
         b(i) = dble(xmp_node_num())
         c(i) = real(xmp_node_num())
      enddo

      sa = 0
      sb = 0.0
      sc = 0.0

!$xmp loop (i) onto t(i) reduction(-: sa)
      do i=1, N
         sa = sa-a(i)
         sb = sb-b(i)
         sc = sc-c(i)
      enddo

      procs = xmp_num_nodes()
      allocate(w(1:procs))
      if(mod(N,procs) .eq. 0) then
         w = N/procs
      else
         do i=1, procs
            if(i .le. mod(N,procs)) then
               w(i) = N/procs+1
            else
               w(i) = N/procs
            endif
         enddo
      endif

      ans = 0
      do i=1, procs
         ans = ans - i*w(i)
      enddo

      result = 'OK'
      if(  sa .ne. ans .or.
     $     abs(sb-dble(ans)) .gt. 0.0000001 .or.
     $     abs(sb-real(ans)) .gt. 0.0001 ) then
         result = 'NG'
      endif

      print *, xmp_node_num(), 'testp154.f ', result
      deallocate(w)

      end

File: tests/XMP/NOT-IMPLEMENTED/288/testp154.c
/*testp154.c*/
/*loop指示文とreduction節(-)のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>  
#include<stdlib.h> 
static const int N=1000;
#pragma xmp nodes p(*)
#pragma xmp template t(0:N-1)
#pragma xmp distribute t(cyclic) onto p
int a[N],sa;
double b[N],sb;
float c[N],sc;
int i,*w;
int ans,procs;
#pragma xmp align a[i] with t(i)
#pragma xmp align b[i] with t(i)
#pragma xmp align c[i] with t(i)
char *result;
int main(void){
#pragma xmp loop (i) on t(i)
   for(i=0;i<N;i++){
      a[i]=xmp_node_num();
      b[i]=(double)xmp_node_num();
      c[i]=(float)xmp_node_num();
   }
   sa = 0;
   sb = 0.0;
   sc = 0.0;

#pragma xmp loop on t(i) reduction(-:sa)
   for(i=0;i<N;i++)
      sa+=a[i];

#pragma xmp loop on t(i) reduction(-:sb)
   for(i=0;i<N;i++)
      sb+=b[i];

#pragma xmp loop on t(i) reduction(-:sc)
   for(i=0;i<N;i++)
      sc+=c[i];

   procs = xmp_num_nodes();
   w = (int *)malloc(procs);
   if(N%procs == 0){
      for(i=1;i<procs+1;i++){
         w[i] = N/procs;
      }
   }else{
      for(i=1;i<procs+1;i++){
         if(i <= N%procs){
            w[i]=N/procs+1;
         }else{
            w[i]=N/procs;
         }
      }
   }
   ans=0;
   for(i=1;i<procs+1;i++){
      ans = ans + i*w[i];
   }
   result="OK";
   if((sa != ans)||abs(sb-(double)ans) >= 0.0000001 ||abs(sc-(float)ans) >= 0.0001){
      result = "NG";  
   }
   printf("%d %s %s\n",xmp_node_num(),"testp154.c",result);
   free(w); 
   return 0;
}
